Implement time picker replacement for DateInput

foxbunny opened the pull request:
- Added new TimePicker component based on the combo box
- Added custom filter feature to ComboBox (and Select)
- Integrated TimePicker to DateInput (for now)
- DateInput was modified to take time in HH:MM format (omitting the unnecessary seconds); this is an internal change only and it still sends the ISO format date/time upstream

I'll note that TimePicker currently emits blank string for both invalid inputs and actual blank input.

I have considered the implication on the time range picker of using the ComboBox in TimePicker. I think there's going to be a need to perhaps scrap ComboBox completely and repurpose it specifically for Time-related stuff, and extract the popup list functionality from it as a generic reusable custom element. That's probably going to be best from the design point of view, but it's too big of a change for this PR. I'm going to rework the time range picker PR along these lines.

foxbunny committed:
Implement time picker replacement for DateInput

foxbunny commented:
It wasn't a rename. GitHub just interpreted it as such. It was actually an accidental Save As. :))

ThiefMaster commented on indico/web/client/js/react/components/time/time_value.js:
```diff
@@ -0,0 +1,226 @@
+   *
+   * See tests for a more complete list of examples.
+   */
+  static fromString(timeOfDayString, format = '24h') {
```
When I type e.g. `23:00` in a field in am/pm mode it gets rejected, instead of being turned into `11:00 PM`

ThiefMaster replied:
Yes, I'd prefer that. Especially when you use a system that's using am/pm as someone who's used to 24h format, it's much more pleasant to be able to type times in a decent format :)

ThiefMaster commented:
I think the latest change may have broken something... it behaves very weirdly in both 12h and 24h mode. FWIW in both cases I tried to type `23:00` but even after the first 1-2 digits it behaves strangely.

![timepicker-bug](https://github.com/user-attachments/assets/5fabd151-cc00-4ae7-b4c9-c48bc1bdc948)

The main culprit appears to be `setInputValue('any');`

FWIW I think on blur (or when selecting something from the dropdown list) the field should set the visible input value to match the configured format, ie when I type `2300` in the field it should become `23:00` or `11:00 pm` depending on the config on blur.

ThiefMaster commented:
Firefox 140, the default date format from my system locale is 24h. It's a normal date/time registration form field with dd/mm/yyyy for the date format and 12h for the time format.

ThiefMaster commented on indico/web/client/js/custom_elements/ind_time_picker.js:
```diff
@@ -0,0 +1,8 @@
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import CustomElementBase from 'indico/custom_elements/_base';
```
This file is pretty much empty at the moment 🙈

foxbunny replied:
Ah crap. Yeah, I was gonna rename ind_combo_box to ind_time_picker and move time-picker-related logic there. Cause we're not really using the combobox for what it was designed for. I can still do this in another PR if you want.

ThiefMaster commented on indico/web/client/js/react/components/time/TimePicker.jsx:
```diff
@@ -0,0 +1,201 @@
+  return optionRank;
+}
+
+export default function TimePicker({value, startTime, step, min, max, timeFormat, onChange}) {
```
I think I forgot to mention it in the initial issue about what this component needs to support, but we currently support disabling specific hours/minutes; this is used in places where we have a "not before a specific date/time" restriction.

I don't think we need it with a full-blown filter callback, being able to set `minTime` and `maxTime` props would be sufficient.

ThiefMaster commented on indico/web/client/js/react/components/time/TimePicker.jsx:
```diff
@@ -0,0 +1,201 @@
+
+  return (
+    <>
+      <ComboBox
```
the `required` prop from the TimePicker needs to be passed to this one so the clear button is not shown in case it's required (I fixed this locally so I'll push a comment for this)

ThiefMaster commented on indico/web/client/js/react/components/time/TimePicker.jsx:
```diff
@@ -0,0 +1,213 @@
+}
+
+export default function TimePicker({
+  value,
```
This seems to be completely ignored; when editing a registration the time picker always shows `--:--`, even though there's a valid time string in this prop.

foxbunny commented on indico/web/client/js/react/components/time/TimePicker.jsx:
```diff
@@ -0,0 +1,201 @@
+  return optionRank;
+}
+
+export default function TimePicker({value, startTime, step, min, max, timeFormat, onChange}) {
```
This comes from how the blur event is handled to close the list. Because mouseup happens *after* blur, if I were to make a selection on mouseup I would have to jump through hoops to do it because blur hits before and closes the list. It's possible but I'd keep it as is if you don't mind. It's a very very edge case.

foxbunny commented on indico/web/client/js/react/components/time/time_value.js:
```diff
@@ -0,0 +1,262 @@
+export function timeList(options = {}) {
+  const {
+    markCurrent = '',
+    startTime = '',
```
Start time is specified separately just for duration calculations. This lets you set minTime without showing duration, or if you want to show duration from a different point than minTime. So you're correct in assuming that you'd set minTime=startTime for the behavior similar to Google calendar. I'll check the step/minTime relationship and make sure you can produce the same kind of offset list of times.

foxbunny commented:
Added time range picker replacement and integrated it into RB as an example. Also updated positioning logic so it works within existing SUI popups. Those popups should still be replaced for other reasons, but I figure it's better to do it in a separate PR.

foxbunny commented:
Sorry for the delay, too. Currently working on the a11y report for the work I've done thus far. I'm gonna look at these issues now.

> Focusing the input field does not auto-select the current content

By this you mean when you "click" on the field, the contents should be selected?

> There's a horizontal scrollbar in the RB end time field

I've meanwhile learned a new trick that might help us isolate styles a bit better. It's, of course, a hack, but it will be much cleaner than that sui-debt thing. It looks like this:

```
:not(#bump #specificity) some-selector {
  .....
}
```

The way it works is the fake `#bump #specificity` id bumps the specificity of the selector to `2 x x` as if we used two `id` selectors (this is technically higher than anything you'd use in real life). In theory, wrapping any of the placeholder selectors in these will bump their specificity to a level where they won't be influenced by surrounding styles. 

The downside is you actually have to know what this means when you first see it (that's why I suggest those ridiculous id names), and that, in order to override these styles you need to use the same hack (can't simply rely on the cascade... although we already can't anyway).

What do you think?

foxbunny commented:
> Focusing the input field does not auto-select the current content

Oh, and would you like the same behavior for the date picker?

ThiefMaster commented:
The `#bump #specificity` hack is certainly a bit of a hack, but both me and @tomasr8 have the feeling that it's not too bad.

I think one way to keep it more maintainable would be mentioning in each place where it's used why exactly it is needed, ie what the selector interferes with otherwise.

ThiefMaster commented on indico/web/client/js/custom_elements/ind_time_picker.js:
```diff
@@ -63,7 +71,24 @@ customElements.define(
+      });
+
       input.addEventListener('click', () => {
+        // Select all text when clicking on the input
+        selectInputText(0);
```
Doing it on focus would have nicer behavior (you can click again to select just parts of it / place the caret).

ThiefMaster commented:
There's another very nice behavior from the Google Calendar time picker which I just discovered:

When you enter a custom time (let's say `13:37`), and the date picker opens, it nonetheless scrolls to the times around it

<img width="239" height="280" alt="image" src="https://github.com/user-attachments/assets/21c9b146-348c-4654-a767-99a17672cabd" />

Same for the end time of course:

<details>
<img width="297" height="276" alt="image" src="https://github.com/user-attachments/assets/b81e0d42-4547-4151-acd9-7706b7ad69c9" />
<img width="304" height="256" alt="image" src="https://github.com/user-attachments/assets/051f7985-a954-477e-bfaf-64aa3f64d53b" />
</details>

foxbunny commented:
> ie what the selector interferes with otherwise.

It's basically a hack to avoid having to think about it. SUI is fundamentally designed to be used by itself on the page, so it doesn't really care about specificity. The new design system is also designed that way. So, in order to have a 'clean win' for the new design system it's better to bump the specificity so high that there's literally no chance of SUI ever interfering (e.g., wrap the entire design system in it, except for the resets). In that case, specifically tracking what selectors interfere would be impossible or very time consuming.

foxbunny commented:
I can't repro the horizontal scrollbar issue. Can you check if it is still happening?

foxbunny commented:
> It would have been nice to just move my test commit(s) into a second commit after your own squashed ones

Sorry, to be honest I forgot it was there until it got flagged by a linter.
